### PR TITLE
Lint for event filter and deprecate `exclude`

### PR DESCRIPTION
**Pull request by qwerty287**
Closes https://github.com/woodpecker-ci/woodpecker/discussions/2174

- return bad habit error if no event filter is set
- If this is applied, it's useless to allow `exclude`s on events. Therefore, deprecate it together with `include`s which should be replaced by `base.StringOrSlice` later.

### CODE REVIEW

**Review by 6543 — CHANGES_REQUESTED** [review]
let me have a look into the discussion first ...

... as i told - discussions are a bad "representation" i normaly dont get either a notice nore the current consensus from it

File: pipeline/errors/error.go [hunk 1]
```diff
@@ -14,6 +14,7 @@ const (
 	PipelineErrorTypeDeprecation PipelineErrorType = "deprecation" // using some deprecated feature
 	PipelineErrorTypeCompiler    PipelineErrorType = "compiler"    // some error with the config semantics
 	PipelineErrorTypeGeneric     PipelineErrorType = "generic"     // some generic error
+	PipelineErrorTypeBadHabit    PipelineErrorType = "bad_habit"   // some generic error
```
Comment (anbraten) [inline pipeline/errors/error.go:17]:
```suggestion
	PipelineErrorTypeBadHabit    PipelineErrorType = "bad_habit"   // some bad-habit error
```

File: docs/docs/91-migrations.md [hunk 2]
```diff
@@ -7,6 +7,7 @@ Some versions need some changes to the server configuration or the pipeline conf
 - Deprecated `steps.[name].group` in favor of `steps.[name].depends_on` (see [workflow syntax](./20-usage/20-workflow-syntax.md#depends_on) to learn how to set dependencies)
 - Removed `WOODPECKER_ROOT_PATH` and `WOODPECKER_ROOT_URL` config variables. Use `WOODPECKER_HOST` with a path instead
 - Pipelines without a config file will now be skipped instead of failing
+- Deprecated `includes` and `excludes` support from event filter
```
Comment (anbraten) [inline docs/docs/91-migrations.md:10]: We might directly remove it from the available docs as well, so no one will start using it.

**Review by anbraten — APPROVED** [review]

**Comment by anbraten** [comment]
@6543 Did you had the chance to read the discussion?

**Comment by 6543** [comment]
Not realy ... and i dont get why we should remove exclude?

**Comment by 6543** [comment]
Thanks github ... i had a detailed text. I recap it in a tldr now: 

When filter currently has AND, OR and NAND, to build all combinations.

removing exclude is removing NAND

There is a workaround: use eval; but we should not encurage long eval statements if the simple filters just work fine

**Review by 6543 — CHANGES_REQUESTED** [review]
PS: mentioned dicsuccion does not argue at all why exclude justd be removed!

**Comment by 6543** [comment]
Ah ok ... well i do use it ... looks like i have to migrate on repo.

Only for events is fine but the codecomment do not indicate that

File: docs/docs/91-migrations.md (hunk 2 above)
Comment (6543) [inline docs/docs/91-migrations.md:10]:
```suggestion
- Deprecated `includes` and `excludes` support from **event** filter
```

File: pipeline/frontend/yaml/linter/linter.go [hunk 3]
```diff
@@ -263,10 +263,86 @@ func (l *Linter) lintDeprecations(config *WorkflowConfig) (err error) {
...
 
+	for i, c := range parsed.When.Constraints {
+		if len(c.Event.Exclude) != 0 {
+			err = multierr.Append(err, &errors.PipelineError{
+				Type:    errors.PipelineErrorTypeDeprecation,
+				Message: "Please only use allow lists for events",
+				Data: errors.DeprecationErrorData{
+					File:  config.File,
+					Field: fmt.Sprintf("when[%d].event", i),
+					Docs:  "https://woodpecker-ci.org/docs/next/usage/workflow-syntax#event-1",
```
Comment (6543) [inline pipeline/frontend/yaml/linter/linter.go:274]:
```suggestion
					Docs:  "https://woodpecker-ci.org/docs/usage/workflow-syntax#event-1",
```

File: pipeline/frontend/yaml/linter/linter.go [hunk 4]
```diff
@@ -263,10 +263,86 @@ func (l *Linter) lintDeprecations(config *WorkflowConfig) (err error) {
...
+	for _, step := range parsed.Steps.ContainerList {
+		for i, c := range step.When.Constraints {
+			if len(c.Event.Exclude) != 0 {
+				err = multierr.Append(err, &errors.PipelineError{
+					Type:    errors.PipelineErrorTypeDeprecation,
+					Message: "Please only use allow lists for events",
+					Data: errors.DeprecationErrorData{
+						File:  config.File,
+						Field: fmt.Sprintf("steps.%s.when[%d].event", step.Name, i),
+						Docs:  "https://woodpecker-ci.org/docs/next/usage/workflow-syntax#event",
```
Comment (6543) [inline pipeline/frontend/yaml/linter/linter.go:290]:
```suggestion
						Docs:  "https://woodpecker-ci.org/docs/usage/workflow-syntax#event",
```

File: pipeline/frontend/yaml/linter/linter_test.go [hunk 5]
```diff
@@ -172,3 +181,41 @@ func TestLintErrors(t *testing.T) {
...
+		from string
+		want string
+	}{
+		{
+			from: "steps: { build: { image: golang } }",
+			want: "Please set an event filter",
+		},
+		{
+			from: "when: [{branch: xyz}, {event: push}]\nsteps: { build: { image: golang } }",
+			want: "Please set an event filter",
```
Comment (6543) [inline pipeline/frontend/yaml/linter/linter_test.go:196]: why does it suggest an event filter if global does already have it?

**Comment by 6543** [comment]
fixed the "code comment" issue ...
... I only found one issue left - but it's a minor thing and if it's hard to fix it now we can let it open for later

File: pipeline/frontend/yaml/linter/linter_test.go (hunk 5 above)
Comment (6543) [reply pipeline/frontend/yaml/linter/linter_test.go]: Well i guess, user having questions abour that message or not will finaly tell us

**Review by 6543 — APPROVED** [review]

**Pull request merged**